lake/root.go

```diff
@@ -83,37 +93,113 @@ func CreateOrOpen(ctx context.Context, engine storage.Engine, path *storage.URI)
 	return Create(ctx, engine, path)
 }
 
-func (r *Root) createConfig(ctx context.Context) (*Config, error) {
+func (r *Root) createConfig(ctx context.Context) error {
 	poolPath := r.path.AppendPath(PoolsTag)
 	rulesPath := r.path.AppendPath(IndexRulesTag)
 	types := []interface{}{PoolConfig{}}
-	pools, err := kvs.Create(ctx, r.engine, poolPath, types)
+	var err error
+	r.pools, err = kvs.Create(ctx, r.engine, poolPath, types)
 	if err != nil {
-		return nil, err
+		return err
 	}
-	indexRules, err := index.CreateStore(ctx, r.engine, rulesPath)
+	r.indexRules, err = index.CreateStore(ctx, r.engine, rulesPath)
 	if err != nil {
-		return nil, err
+		return err
 	}
-	return &Config{pools, indexRules}, nil
+	if err := r.writeLakeMagic(ctx); err != nil {
+		return err
+	}
+	return err
 }
 
 func (r *Root) loadConfig(ctx context.Context) error {
+	if err := r.readLakeMagic(ctx); err != nil {
+		return err
+	}
 	poolPath := r.path.AppendPath(PoolsTag)
 	rulesPath := r.path.AppendPath(IndexRulesTag)
 	types := []interface{}{PoolConfig{}}
-	pools, err := kvs.Open(ctx, r.engine, poolPath, types)
+	var err error
+	r.pools, err = kvs.Open(ctx, r.engine, poolPath, types)
 	if err != nil {
 		return err
 	}
-	indexRules, err := index.OpenStore(ctx, r.engine, rulesPath)
+	r.indexRules, err = index.OpenStore(ctx, r.engine, rulesPath)
+	return err
+}
+
+func (r *Root) writeLakeMagic(ctx context.Context) error {
+	if err := r.readLakeMagic(ctx); err == nil {
+		return errors.New("lake already exists")
+	}
+	magic := &LakeMagic{
+		Magic:   LakeMagicString,
+		Version: Version,
+	}
+	serializer := zngbytes.NewSerializer()
+	if err := serializer.Write(magic); err != nil {
+		return err
+	}
+	if err := serializer.Close(); err != nil {
+		return err
+	}
+	path := r.path.AppendPath(LakeMagicFile)
+	err := r.engine.PutIfNotExists(ctx, path, serializer.Bytes())
+	if err == storage.ErrNotSupported {
+		//XXX workaround for now: see issue #2686
+		reader := bytes.NewReader(serializer.Bytes())
+		err = storage.Put(ctx, r.engine, path, reader)
+	}
+	return err
+}
+
+func (r *Root) readLakeMagic(ctx context.Context) error {
+	path := r.path.AppendPath(LakeMagicFile)
+	reader, err := r.engine.Get(ctx, path)
 	if err != nil {
 		return err
 	}
-	r.Config = &Config{pools, indexRules}
+	deserializer := zngbytes.NewDeserializer(reader, []interface{}{
+		LakeMagic{},
+	})
+	v, err := deserializer.Read()
+	if err != nil {
+		return err
+	}
+	magic, ok := v.(*LakeMagic)
+	if !ok {
+		return fmt.Errorf("corrupt lake magic file %q: unknown type: %T", LakeMagicFile, v)
+	}
+	if magic.Magic != LakeMagicString {
+		return fmt.Errorf("corrupt lake magic: %q should be %q", magic.Magic, LakeMagicString)
```

Comment:
```suggestion
		return fmt.Errorf("corrupt lake magic file: %q should be %q", magic.Magic, LakeMagicString)
```

Reply:
This is the magic field inside of the file.  We can call the file the version file and refer to the magic inside the file.